Use a strict record type for `ProtocolParameters`.

jonathanknowles opened the pull request:
## Issue

ADP-2903

## Summary

This PR replaces the use of a lazily-evaluated tuple with a strict `ProtocolParameters` record type.

## Details

The goal is to prevent repeated evaluation of the `BundledProtocolParameters` value in loops that may suffer from loss of sharing when compiled with all optimisations enabled (i.e., when compiled with `-O2`).

Since protocol parameters are "read" from the blockchain, we locate the `ProtocolParameters` type within the `Read` module hierarchy instead of the `Write` hierarchy.

## Future work

This PR does not make the `ProtocolParameters` type abstract in any way: the default constructor is currently exported. We can revise this in future, if and when necessary.

jonathanknowles committed:
Use a strict record type for protocol parameters.

This replaces the use of a lazy 2-tuple.

jonathanknowles committed:
Generalise `toBalanceTxPParams` to any Shelley-based era.

jonathanknowles committed:
Rename `toBalanceTxPParams` to `unsafeFromWalletProtocolParameters`.

jonathanknowles committed:
Move `unsafeFromWalletProtocolParameters` to `Read.ProtocolParameters`.

jonathanknowles committed:
Improve readability of `unsafeFromWalletProtocolParameters`.

jonathanknowles committed:
Add reminder to make `ProtocolParameters` data type abstract.

Anviking commented on lib/wallet/src/Cardano/Wallet/Read/ProtocolParameters.hs:
```diff
@@ -0,0 +1,45 @@
+--
+-- Provides protocol parameters.
+--
+module Cardano.Wallet.Read.ProtocolParameters
```
Shouldn't this be in `Write`? Possibly just in `Write.Tx.Balance` even?

jonathanknowles replied:
> Shouldn't this be in `Write`?

Since protocol parameters are generally "read" from the blockchain, I simply chose to locate the `ProtocolParameters` type within the `Read` module hierarchy instead of the `Write` hierarchy.

jonathanknowles replied:
I should add that this "version" of the `ProtocolParameters` type is not in any way intended to be the final version of the type we would ultimately like to see in the `Read`/`Write` hierarchy.

The intention of this PR is solely to fix the current performance issue, and unblock the release.

In follow-up PRs, we are still free to do any or all of the following:
- move the type somewhere else
- make the type abstract
- change the type definition

HeinrichApfelmus replied:
> Since protocol parameters are generally "read" from the blockchain, 

Yes. As the purpose of the `Cardano.Wallet.Read` hierarchy is to provide Haskell types and functions for stuff that appears on the blockchain, it is within the purview of the `Read` hierarchy to contain a module for `ProtocolParameters`. This can be the beginning of that module.

In contrast, the purpose of the `Cardano.Wallet.Write` hierarchy is to provide Haskell types and functions for stuff that is accepted into the current and subsequent era of the blockchain. Sometimes, no equivalent in `Read` is needed, but often, `Write` is a selected subset of `Read`. Hence, a typical case is two modules:

* `Cardano.Wallet.Read.ProtocolParameters`
* `Cardano.Wallet.Write.ProtocolParameters` — re-exporting a subset of the `Cardano.Wallet.Read.ProtocolParameters` with specialized types.

Can we get away with having `ProtocolParameters` in `Write` only? I would argue that the answer is "no". I completely agree that for making a transaction, we only need the latest protocol parameters, so that's a good argument in favor of "yes". However, the problem is that if the `cardano-node` is not fully synched, it might give us protocol parameters from old eras — and we need to be able to read those, e.g. for converting slot numbers to wall-clock time.

In that sense, it's fair game for `Read` to have a type `ProtocolParameters` for each era of the Cardano blockchain as they are defined in the specification. For instance, here is the protocol parameters specification for the Shelley era:

<img width="478" alt="PParams Shelley" src="https://user-images.githubusercontent.com/564592/231725507-39c4fe68-19cc-4c35-a2a8-641581e38413.png">

HeinrichApfelmus replied:
In short, I'm actually in favor of *two* modules:

* `Cardano.Wallet.Read.ProtocolParameters`
* and `Cardano.Wallet.Write.ProtocolParameters`, which re-exports the previous one.

In order to preserve abstraction boundaries, modules in the (vicinity of the) `Write` hierarchy should import the `Cardano.Wallet.Write.ProtocolParameters`; this applies to many files in this pull request, see one comment.

HeinrichApfelmus replied:
After discussing with Johannes, we came to the following conclusion:

* The type introduced in this pull request should live in `Cardano.Wallet.Write.ProtocolParameters`.

Furthermore:
* The `Cardano.Wallet.Write` hierarchy need not depend on `Cardano.Wallet.Read`, and can depend on ledger types directly.
  * I see `Cardano.Wallet.Read` as a way to represent concepts from the ledger specification of each era as Haskell types — i.e. I see it as a replacement for `cardano-ledger` (that also happens to be implemented using `cardano-ledger`). In this way, `Read` can absorb changes in the `cardano-ledger` code that are not reflected in the specification. This also implies that I find it natural to have `Write` depend on `Read`.
  * However, Johannes prefers `Write` to depend on `cardano-ledger` rather than `Read` because
    * Fee evaluation operates on types from `cardano-ledger`.
    * The dependency footprint for users of the `Write` hierarchy is reduced.
  * I don't object to the decision of having `Write` *not* depend on `Read` — as long as the abstraction "`Write` is about writing stuff in the current or subsequent era" is preserved, I would consider this dependency an implementation detail.
* We can't think of a good use case for `Cardano.Wallet.Read.ProtocolParameters` — old parameters may arise when the node is not fully synchronized, but the most "interesting" thing that the wallet can do in that case is to refuse making transactions.

jonathanknowles replied:
@HeinrichApfelmus, @Anviking

Thank-you both for looking at this.

I've added 70e564a99f81f1fc96696fa62f8d494754397ecf which moves the newly-added `ProtocolParameters` module from the `Read` hierarchy to the `Write` hierarchy.

HeinrichApfelmus commented on lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs:
```diff
@@ -76,6 +76,8 @@ import Cardano.Wallet.Primitive.Types.UTxOIndex
     ( UTxOIndex )
 import Cardano.Wallet.Primitive.Types.UTxOSelection
     ( UTxOSelection )
+import Cardano.Wallet.Read.ProtocolParameters
```
```suggestion
import Cardano.Wallet.Write.ProtocolParameters
```

jonathanknowles replied:
Thanks @HeinrichApfelmus: this is fixed in 70e564a99f81f1fc96696fa62f8d494754397ecf.

jonathanknowles committed:
Move `Read.ProtocolParameters` to `Write.ProtocolParameters`.

This commit moves the recently-introduced `ProtocolParameters` module
from the `Read` module hierarchy to the `Write` module hierarchy.

Anviking commented on lib/wallet/src/Cardano/Wallet/Write/ProtocolParameters.hs:
```diff
@@ -0,0 +1,45 @@
+import qualified Cardano.Api as CardanoApi
+import qualified Cardano.Wallet.Primitive.Types as Wallet
+
+-- TODO: Make this data type abstract: don't export the constructor.
```
"TODO: Replace this type with a re-exported `Ledger.PParams era`" 😛

jonathanknowles replied:
@Anviking done in 92977d7858ff46350c421b5d1320776fc94f4275.

jonathanknowles committed:
Add a reminder to re-define `ProtocolParameters` in terms of ledger params.

In response to review feedback:
https://github.com/input-output-hk/cardano-wallet/pull/3851#discussion_r1165445124

Anviking reviewed (approved):
Thanks!
